### PR TITLE
Parallelize `gatherTopK` on multiple blocks

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This PR adds `mbtopk::gatherTopK` which uses the same number of blocks as `radixFindKthValues` to gather top k values in parallel. With this new `gatherTopK` kernel, the sort path is no longer needed because `best(mb, sb)` now is always better than the sort path.

## Algorithm
During each pass of `radixFindKthValues`, this kernel will output per slice `desired` and per block digit counters. After every pass, I use kernel `computeBlockwiseWithinKCounts` to compute the number of elements that `>kthValue`(if largest) or `<kthValue`(if !largest)  for each block from the per slice `desired` and per block digit counters. After the last pass, I use ~kernel `computeBlockwiseKthCounts`~ (~edit: fancy iterator~ edit 2: fancy iterator is too large in binary size, so I decided to use `computeBlockwiseKthCounts` in the end) to compute the number of elements that `==kthValue` for each block from the per slice `desired` and per block digit counters. Then I use cub's scan-by-key algorithm to compute the indices in output where each block should write its output to. Then I used the `mbtopk::gatherTopK` to write top k elements to these indices.

## Benchmark

Using script from @yueyericardo: https://github.com/yueyericardo/misc/blob/master/share/topk-script/benchmark.py

I get the following result on RTX3090:

**New mb vs old mb speedup:**
plot_new

**New dispatched vs old dispatched speedup:**
plot_new-dispatched

Raw data in: https://docs.google.com/spreadsheets/d/e/2PACX-1vQxECU_qP1G-skQ8DmJDo-hx0OYPiJ01EkJMSZQdzfWj-QxhScQCkj9Z3KKBc7svEA6DC03UNvD1ial/pubhtml
This spreadsheet shows that the sort path is no longer needed: `best(mb, sb)` now is always better than the sort path.

cc: @yueyericardo @ngimel @ptrblck

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
Some additional benchmark:

**New mb vs sb speedup:**
plot_new

**New mb vs sort speedup:**
plot_new

**Comment by ngimel** [comment]
What's the context size hit? Also, which cuda version is needed for this, 11.6? What else do we need 11.6 for, some embedding things?

**Comment by zasdfgbnm** [comment]
Yes, it is CUDA 11.6. And Yes, the scan by key in embedding also needs CUDA 11.6. But fully removal of sync in Embedding needs unique by key, which requires higher version of CUDA.

**Comment by zasdfgbnm** [comment]
Binary size 2.8MB -> 8.8MB :(

**Comment by zasdfgbnm** [comment]
Probably I can add a constraint that k must be <= INT_MAX. I think this is a valid constraint.

**Comment by zasdfgbnm** [comment]
Reverting my last commit [Use fancy iterator for kth counts](https://github.com/pytorch/pytorch/pull/74267/commits/d9ca9121d44198acf44ea64c168c6c4c38b0cfcf) reduces the binary size from 8.8MB to 4.3MB. So I would blame cub for being large :(

**Comment by zasdfgbnm** [comment]
For the remaining 4.3MB, cub `inclusive_sum_by_key` takes 1MB, `gatherTopK` takes 0.9MB, while the previous `gatherTopK`  only takes 0.4MB.

**Comment by zasdfgbnm** [comment]
@ngimel I reduced the binary size to 3.1MB, before this PR, it was 2.8MB. So hopefully this is fine.

**Comment by zasdfgbnm** [comment]
Interestingly, after tunning the binary size, ther perf improved a lot:

**New dispatched vs old dispatched speedup before my binary size reduction:**
plot_new-dispatched
**New dispatched vs old dispatched speedup after my binary size reduction:**
plot_new-dispatched

**Comment by ngimel** [comment]
Nice! Was IndexT somehow set to int64 by mistake?

**Comment by zasdfgbnm** [comment]
I don't think it was `IndexT` being accidentally set to `int64_t`. My guess is because I changed the type of per block digit counter in global memory from `IndexT` to short, which reduces memory traffic for the cross-block reduction of `radixFindKthValues` and `computeBlockwiseWithinKCounts`.

**Comment by zasdfgbnm** [comment]
ping @ngimel for review :)

**Comment by ngimel** [comment]
Should we also retire kthvalue and make it always call topk, since it looks like it'll always be slower?

File: aten/src/ATen/native/cuda/TensorTopK.cu [hunk 1]
```diff
@@ -429,20 +652,31 @@ void launch(
...
+  AT_CUDA_CHECK(cudaMemsetAsync(semaphores, 0, numInputSlices * sizeof(uint32_t), stream));
 
-  auto ks_to_find_buffer = allocator.allocate(numInputSlices * sizeof(IndexType));
-  IndexType* ks_to_find = reinterpret_cast<IndexType*>(ks_to_find_buffer.get());
-  IndexType k_to_find = largest ? inputSliceSize - outputSliceSize + 1: outputSliceSize;
-  fill<IndexType><<<std::min((numInputSlices + 511) / 512, (IndexType)65535), 512, 0, c10::cuda::getCurrentCUDAStream()>>>(
+  auto ks_to_find_buffer = allocator.allocate(numInputSlices * sizeof(uint32_t));
+  uint32_t* ks_to_find = reinterpret_cast<uint32_t*>(ks_to_find_buffer.get());
+  uint32_t k_to_find = largest ? inputSliceSize - outputSliceSize + 1: outputSliceSize;
+  fill<uint32_t><<<std::min((numInputSlices + 511) / 512, (uint32_t)65535), 512, 0, stream>>>(
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/TensorTopK.cu:660]: nit - `min` is not needed here, blockDim.x can be 2**31?

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/native/cuda/TensorTopK.cu (hunk 1 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/TensorTopK.cu]: I think the largest value is `2**31-1`. For this line, probably the best is to use
```C++
std::min(((int64_t)numInputSlices + 511) / 512, (int64_t)some_large_number)
```
to handle overflow. I updated this PR with it.

File: aten/src/ATen/native/cuda/TensorTopK.cu [hunk 2]
```diff
@@ -709,7 +709,7 @@ void launch(
   }
 
 #if CUB_SUPPORTS_SCAN_BY_KEY()
-  computeBlockwiseKthCounts<Bitwise><<<std::min((numInputSlices + 255) / 256, (uint32_t)65535), 256, 0, stream>>>(
+  computeBlockwiseKthCounts<Bitwise><<<std::min(((int64_t)numInputSlices + 255) / 256, (int64_t)1073741824), 256, 0, stream>>>(
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/TensorTopK.cu:712]: Well, realistically memory that's enough to hold 256 G slices won't appear in our lifetimes, but ok.

**Pull request closed without merging**